src/persist-client/src/dyn_cfg.rs

```diff
@@ -0,0 +1,409 @@
+//!   config is used in only one place and this makes it easy to see the
+//!   associated documentation at the usage site. Configs that are used in
+//!   multiple places may be defined in some common place as appropriate.
+//! - Everything is type-safe.
+//! - Secondarily: set up the ability to get and use the latest values of
+//!   configs in tooling like `persistcli` and `stash debug`. As we've embraced
+//!   dynamic configuration, we've ended up in a situation where it's stressful
+//!   to run the read-write `persistcli admin` tooling with the defaults
+//!   compiled into code, but `persistcli` doesn't have access to the vars stuff
+//!   and doesn't want to instantiate a catalog impl.
+
+use std::collections::BTreeMap;
+use std::sync::atomic::{AtomicBool, AtomicU64};
+use std::sync::{Arc, RwLock};
+
+use tracing::error;
+
+use mz_proto::{ProtoType, RustType};
+
+include!(concat!(env!("OUT_DIR"), "/mz_persist_client.dyn_cfg.rs"));
+
+/// A handle to a dynamically updatable configuration value.
+///
+/// This represents a strongly-typed named config of type `T`. It may be
+/// registered to a set of such configs with [ConfigSet::add] and then later
+/// used to retrieve the latest value at any time with [Self::get].
+///
+/// The supported types are [bool], [usize], and [String].
+///
+/// TODO(cfg): Add Duration in a followup.
+#[derive(Clone, Debug)]
+pub struct Config<T> {
+    name: &'static str,
+    desc: &'static str,
+    default: T,
+}
+
+impl<T: ConfigType> Config<T> {
+    /// Constructs a handle for a config of type `T`.
+    ///
+    /// It is best practice, but not strictly required, for the name to be
+    /// globally unique within a process.
+    pub const fn new(name: &'static str, default: T, desc: &'static str) -> Self {
+        Config {
+            name,
+            default,
+            desc,
+        }
+    }
+
+    /// The name of this config.
+    pub fn name(&self) -> &str {
+        self.name
+    }
+
+    /// The description of this config.
+    pub fn desc(&self) -> &str {
+        self.desc
+    }
+
+    /// The default value of this config.
+    pub fn default(&self) -> &T {
+        &self.default
+    }
+
+    /// Adds this config to the given set.
+    ///
+    /// Names are required to be unique within a set, but each set is entirely
+    /// independent. The same `Config` may be registered to multiple
+    /// [ConfigSet]s and thus have independent values (e.g. imagine a units test
+    /// executing concurrently in the same process).
+    ///
+    /// Panics if a config with the same name has previously been registered to
+    /// this set.
+    pub fn register(&self, set: &mut ConfigSet) -> T {
+        T::get(T::shared(self, set).expect("config should be registered to set"))
+    }
+
+    /// Returns the latest value of this config within the given set.
+    ///
+    /// Panics if this config was not previously registered to the set.
+    pub fn get(&self, set: &ConfigSet) -> T {
```

Comment:
I did this because when you're getting a config value, the critical "noun" is more which config than the bag of values. Concretely, I find

```
if STREAMING_SNAPSHOT_AND_FETCH_ENABLED.get(&self.machine.applier.cfg.configs) {
```

easier to read than

```
if self.machine.applier.cfg.configs.get(&STREAMING_SNAPSHOT_AND_FETCH_ENABLED) {
```

Secondarily, I _think_ rustfmt might do a worse job on the latter (break it up into a bunch of lines and make it hard to follow the control flow).

That said, not a hill I'd die on! I'm curious if anyone else on this PR has a strong opinion one way or the other?